horovod/spark/keras/datamodule.py

```diff
@@ -0,0 +1,169 @@
+                                storage_options=self.storage_options,
+                                # Don't shuffle row groups if shuffle_buffer_size is 0 (non-shuffle case).
+                                shuffle_row_groups=True if self.shuffle_size > 0 else False,
+                                **self.reader_factory_kwargs)
+        self.val_reader = self.reader_factory(self.val_dir,
+                                num_epochs=1,
+                                cur_shard=hvd.rank(),
+                                reader_pool_type=self.reader_pool_type,
+                                workers_count=self.val_reader_worker_count,
+                                shard_count=hvd.size(),
+                                hdfs_driver=constants.PETASTORM_HDFS_DRIVER,
+                                schema_fields=self.schema_fields,
+                                transform_spec=self.transform_spec,
+                                storage_options=self.storage_options,
+                                shuffle_row_groups=False,
+                                **self.reader_factory_kwargs) if self.has_val else self.empty_batch_reader()
+        return self
+
+    def __exit__(self, type, value, traceback):
+        if self.has_val and self.val_reader:
+            self.val_reader.stop()
+        if self.train_reader:
+            self.train_reader.stop()
+        super().__exit__(type, value, traceback)
+
+    @contextlib.contextmanager
+    def empty_batch_reader(self):
+        yield None
+
+    def train_data(self):
+        return self.make_dataset(self.train_reader, self.train_batch_size, self.shuffle_size,
+                                 self.is_batch_reader, shuffle=True if self.shuffle_size > 0 else False,
+                                 cache=self.inmemory_cache_all, seed=self.random_seed)
+
+    def val_data(self):
+        return self.make_dataset(self.val_reader, self.val_batch_size, self.shuffle_size,
+                                 self.is_batch_reader, shuffle=False, cache=self.inmemory_cache_all) if self.val_reader else None
+
+
+
+class NVTabularDataModule(DataModule):
+    """NVTabular-based DataModule for KerasEstimator for GPU-accelerated data loading of tabular datasets.
+
+    Note: requires `label_cols`, `categorical_cols`, and `continuous_cols` to be explicitly provided."""
+    def __init__(self, label_cols=[], categorical_cols=[], continuous_cols=[], **kwargs):
+        super().__init__(**kwargs)
+        self.label_cols = label_cols
+        self.categorical_cols = categorical_cols
+        self.continuous_cols = continuous_cols
+
+    @staticmethod
+    def seed_fn():
+        """
+        Generate consistent dataloader shuffle seeds across workers
+        Reseeds each worker's dataloader each epoch to get fresh a shuffle
+        that's consistent across workers.
+        """
+        import horovod.tensorflow.keras as hvd
+        import cupy
```

Comment:
Probably can get rid of this particular dependency, since it's just generating a random number.  Will test replacing w/ alternatives.

Reply:
Replaced `cupy` with `numpy` here, since this is just generating a random number.  As such, we probably don't want to add `nvtabular` as an `extras_require`, since it's unlikely to successfully install via `pip install`.  Instead, we can refer folks to the `docker/horovod-nvtabular` image.